misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -231,6 +253,71 @@ internal class SqsJobConsumer @Inject internal constructor(
         )
       }
     }
+
+  }
+
+  internal inner class BatchQueueReceiver(
+    queueName: QueueName,
+    private val handler: BatchJobHandler
+  ) : QueueReceiver(queueName) {
+
+    override fun receive(): List<Status> {
+      // Repeatedly long poll for messages until we have batchSize messages.
+      // If there are no messages available, short circuit further requests.
+      val cutoff = TimeSource.Monotonic.markNow() + 20.seconds // TODO make configurable
+      val batch: MutableList<SqsJob> = mutableListOf()
+      while (batch.size < batchSize() && !cutoff.hasPassedNow()) {
```

Comment:
There will be `SQS_CONSUMERS_PER_POD` threads doing this long poll, waiting until they get the desired batch size, and then each thread will handle a batch of that size. What if we override `run()` in `BatchQueueReceiver` to instead have `SQS_CONSUMERS_PER_POD` poll and combine their results to reach `batchSize()` more quickly? 

I'm not sure how surprising that would be to a user, I'm mostly thinking that for largish batch sizes, it will take many RPCs to reach the threshold, and making the threads build the batch cooperatively could alleviate that

Reply:
I was thinking of the case where the desired batch size is "large" compared to SQS's enforced limit of 10 messages per call, and there are many messages in the queue (e.g. my service has a few million messages it needs to process some days). In that case, there wouldn't be contention among the threads to get enough jobs, so the threads could cooperatively build a batch. However, perhaps it doesn't matter in this case, since there will be enough jobs to hydrate each thread's batch, and it's probably an anti-pattern to try to build a batch that's 10x or more the SQS read batch limit. 

In the case where messages are sparse though, I think it's suboptimal to have multiple threads consuming on the same pod. In the worst case, their batches would starve each other and always end up waiting `cutoff` seconds, decreasing both latency and throughput. I think that could just be called out in documentation though (i.e. if a service wants to batch process messages from a queue with low messages per sec it should reduce SQS_CONSUMERS_PER_POD to 1)